modules/strings/strings.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package strings
+
+import (
+	"strings"
+	"unicode/utf8"
+)
+
+// ToASCIIUpper returns s with all ASCII letters mapped to their upper case.
+func ToASCIIUpper(s string) string {
+	isASCII, hasLower := true, false
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if c >= utf8.RuneSelf {
+			isASCII = false
+			break
+		}
+		hasLower = hasLower || ('a' <= c && c <= 'z')
+	}
+
+	// Optimize for ASCII-only strings.
+	if isASCII {
```

Comment:
IIRC, UTF-8 is ASCII-safe, so we could do `if 'a' <= c && c <= 'z' { c -= 'a' - 'A' }` on every byte safely directly.

Reply:
Yes, we could de-duplicate but the overhead from utf8.Decode is noticeable.

Reply:
We do not need utf8.Decode actually. We can just do `if 'a' <= c && c <= 'z' { c -= 'a' - 'A' } ` for every byte, it's fastest.

I believe the code also pass the test cases:

```go
func ToUpperASCII(s string) string {
    b := []byte(s)
    for i, c := range b {
        if 'a' <= c && c <= 'z' { b[i] -= 'a' - 'A' }
    }
    return string(b)
}
```

Reply:
And maybe we can just put the code to `modules/util` package. The customized `strings` package conflicts with official `strings` package, alias is a little annoying.

Reply:
We might need a new test-case then, because that code will not work on strings with certain utf8 characters, those characters can be made out of multiple bytes, whereby one of those bytes could pass the `'a' <= c && c <= 'z'` but they shouldn't.

Reply:
> We might need a new test-case then, because that code will not work on strings with certain utf8 characters, those characters can be made out of multiple bytes, whereby one of those bytes could pass the `'a' <= c && c <= 'z'` but they shouldn't.

It does work. One goal of UTF-8 design is: never break ASCII string functions. So it's safe to do it.

Reply:
From the table, we can see that 'a' to 'z' are not used by UTF-8 encoding

![image](https://user-images.githubusercontent.com/2114189/151653754-60efba88-231f-48a1-9763-4679a1212516.png)